### PR TITLE
Remaining binary operations

### LINKED ISSUE #246: Add remaining binary operations for Dataset and its proxy classes
Follow up to #247.

- Add `-=` and `/=` for `Dataset`, `DataProxy`, and `DatasetProxy`.
- Add `+`, `-`, `*`, and `/`for the above. Note that the variants where both operands are `DataProxy` do not exist for now (see #245).
- As a bonus, consider ways to avoid copies, e.g., by support rvalue arguments in case of `Dataset`.

### PR DESCRIPTION
**Pull request by DanNixon**
Adds `-=` and `/=` operators for `Dataset`, `DataProxy`, and `DatasetProxy`.

Fixes partially #246.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
I think github will still close the issue if you write "partially fixes #xxx" in the PR message?

**Comment by DanNixon** [comment]
> What about the non-in-place operators?

We discussed this on Slack and concluded they were not a high priority right now.

**Comment by SimonHeybrock** [comment]
> > What about the non-in-place operators?
> 
> We discussed this on Slack and concluded they were not a high priority right now.

I though we agreed that having, e.g., `operator+` with an rvalue or value argument are not a priority. We need those with lvalue references of course.

**Comment by SimonHeybrock** [comment]
> I'm not clear on `apply_with_delay` but otherwise looks good.

It is to avoid issues with self-overlap, e.g., when you do `dataset -= dataset['a']`. Operating on the `'a'` item in the set has to be done last.

**Comment by DanNixon** [comment]
> > > What about the non-in-place operators?
> > 
> > 
> > We discussed this on Slack and concluded they were not a high priority right now.
> 
> I though we agreed that having, e.g., `operator+` with an rvalue or value argument are not a priority. We need those with lvalue references of course.

Ah, sorry I misunderstood in that case. I'll add those too.

File: core/dataset.cpp [hunk 1]
```diff
@@ -680,4 +744,192 @@ std::ostream &operator<<(std::ostream &os, const Dim dim) {
...
+
+Dataset operator+(Dataset &&lhs, const DatasetConstProxy &rhs) {
+  return apply(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(Dataset &&lhs, const DataConstProxy &rhs) {
+  return apply_with_delay(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(const Dataset &lhs, Dataset &&rhs) {
```
Comment (OwenArnold) [inline core/dataset.cpp]: I would expect `operator-` had an overload like this too.

Comment (DanNixon) [reply core/dataset.cpp]: I don't think there is an easy way to implement this based on what operations are already available.

Comment (OwenArnold) [reply core/dataset.cpp]: I think based on non-commutative behaviour, this cannot be implemented, as you say.

File: core/dataset.cpp [hunk 2]
```diff
@@ -680,4 +755,192 @@ std::ostream &operator<<(std::ostream &os, const Dim dim) {
...
+Dataset operator*(Dataset &&lhs, const DatasetConstProxy &rhs) {
+  return apply(times_equals, lhs, rhs);
+}
+
+Dataset operator*(Dataset &&lhs, const DataConstProxy &rhs) {
+  return apply_with_delay(times_equals, lhs, rhs);
+}
+
+Dataset operator*(Dataset &&lhs, Dataset &&rhs) {
+  return apply(times_equals, lhs, rhs);
```
Comment (SimonHeybrock) [inline core/dataset.cpp]: I do not quite understand this and other similar other overloads. What is the point of passing both arguments by rvalue reference? And even if we have just one rvalue ref, does it not need to be moved to have any effect? Furthermore, `apply` accepts arguments by reference (as far I I can see), so even if we move here, we would still need to adapt `apply`?

Comment (DanNixon) [reply core/dataset.cpp]: Overloads where both arguments are rvalue references are needed in the case when you have both rvalue - lvalue and lvalue - rvalue pairs, which are ambiguous for rvalue - ravlue.

This is only actually needed for `operator+` however.

Comment (SimonHeybrock) [reply core/dataset.cpp]: But as long as you have no code that determines which one to move from I do not see the point. Just use const lvalue ref then?

Comment (DanNixon) [reply core/dataset.cpp]: Yes, I've removed it for all but `operator+` where it is required.

Comment (SimonHeybrock) [reply core/dataset.cpp]: Can you elaborate, why is it required for `+`?

Comment (DanNixon) [reply core/dataset.cpp]: Cause `operator+` has (among others) the overloads:
```cpp
Dataset &&operator+(Dataset &&lhs, const Dataset &rhs);
Dataset &&operator+(const Dataset &lhs, Dataset &&rhs);
```

Without `Dataset &&operator+(Dataset &&lhs, Dataset &&rhs);` this operator is ambiguous when both sides are rvalues.

Comment (SimonHeybrock) [reply core/dataset.cpp]: Why only for `+`?

In any case, didn't we conclude in a slack discussion a while ago that having overloads with rvalue ref arguments is too much work and possibly not so useful?

**Pull request merged**